### DISCUSSION TITLE
Pipeline depth calculation

### DISCUSSION THREAD

**Discussion by suarezvictor** [JulianKemmerer/PipelineC#17] [Ideas]
What's the algorithm for fmax/depth estimation?
I mean, does "coarse" latency tests advances one by one until maximum frequency is achieved? if that's the case, I think a successive approximation algorithm would perform better (i.e. to find optimum latency between 1 a 1024, you first test 512, if ok test midpoint at 256 if not, midpoint between 512 and 1024 and continuing recursively)

**Comment by JulianKemmerer** [comment]
@suarezvictor Good question - this isnt really documented anywhere other than a little in the -h

```
$ ./src/pipelinec -h
usage: pipelinec [-h] [--out_dir OUT_DIR] [--sim_comb] [--sim] [--edaplay] [--mult MULT] [--coarse] [--start START] [--sweep] [c_file]

PipelineC:

positional arguments:
  c_file             C file containing main functions. Default is /media/1TB/Dropbox/PipelineC/git/PipelineC/main.c.

optional arguments:
  -h, --help         show this help message and exit
  --out_dir OUT_DIR  Output directory. Default is a new directory inside the current directory.
  --sim_comb         Start simulation using just comb. logic versions of functions.
  --sim              Start simulation using final (maybe autopipelined) versions of functions.
  --edaplay          Setup simulation files for use on edaplayground.com.
  --mult MULT        infer|fabric : Global multiplier style. Default to 'inferring' multiplier DSP primitives. Alternatively implement multipliers in FPGA fabric.
  --coarse           Do only a single coarse grain autopipelining sweep (as opposed to default of several multi-level coarse sweeps).
  --start START      When doing a --coarse sweep, start from N cycles latency.
  --sweep            When doing a --coarse sweep, sweep clock by clock upwards increasing latency.
```

When you run the tool `--coarse` on `rsqrtf.c` you will see this message printed 
```
Main Function: llvm_dis_Z11float_rsqrtf does not have a set target frequency.
Starting a coarse sweep incrementally from zero latency comb. logic.
```

That is the tool telling you it has inferred essentially a `--start 0` and `--sweep`. This is because as checked into the repo there is no specified target frequency (no `MAIN_MHZ` pragma). So the tool, not knowing what fmax/~latency you want, defaults to starting from 0 clocks and just increments from there until it crashes or you kill it...

To get the tool to actually complete it has to know it has met timing. So you have to `specify an fmax goal`. I.e. see `MAIN_MHZ` pragmas near function definition in file. Then you can run with `--coarse` and instead the tool will do like so:

```
loop:
    run pnr
    if not met timing:
       adjust pipelining based on timing results of prev run
    else:
       done
```
So you should see the tool no longer start at 0 clocks, and should increment more quickly trying to reach the target fmax specifically (as opposed to just sweeping latency).

This is definitely open to lots of parallelization and doing things like 'looking ahead' and doing a binary search of like an entire `between 1 a 1024` latency range or something.

Finally, try running `without --coarse` option (with fmax specified still) and see if you like those results. The default mode (not coarse) does a deeper pipeline of submodules modules that tends to work better for larger designs/really high fmax but - we'll see!

**Comment by JulianKemmerer** [comment]
There a good few internal params that could be tuned too - definitely open to adding any more knobs/options for users
